<img align="right" src="https://ds-cs-images.s3.ap-northeast-2.amazonaws.com/Codestates_Fulllogo_Color.png" width=100>

## *AIB / SECTION 2 / SPRINT 3 / NOTE 2*

# 📝 Assignment
---
# Data Wrangling


### 1) 지속적으로 여러분의 포트폴리오 데이터를 랭글링 하세요
- 여러분이 선택한 데이터들을 랭글링하여 모델을 만들기 위한 테이블 형태의 데이터로 만들어 보세요.
- **(Urclass Quiz) 다음 함수의 result를 완성시키고 리턴값을 제출하세요.**

In [ ]:
### 이곳에서 Urlcass 퀴즈(assignment)를 진행해 주세요 ###
import pandas as pd
df = pd.read_csv("https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/assignments/purchase_data.csv", sep=",",index_col=0)

## 다음 result 함수 안에 male, female 변수를 계산하세요.(groupby, get_group을 사용하세요.)
## sum_male: Age = 20, Gender = "Male" 그룹의 Price를 합하세요.
## sum_female: Age = 20, Gender = "Female" 그룹의 Price를 합하세요.

def result(df):

    # Age가 20인 데이터중
    # Groupby로 gender별 price를 sum 해서 
    # Male, Female인 것만 추출
    sum_male = df[df['Age'] == 20].groupby('Gender')['Price'].sum().loc['Male']
    sum_female = df[df['Age'] == 20].groupby('Gender')['Price'].sum().loc['Female']

    print("sum_male: {}, sum_female: {}".format(sum_male, sum_female))
    return round(sum_male + sum_female)

result(df)

sum_male: 260.2, sum_female: 43.730000000000004


304

In [ ]:
df.columns

Index(['SN', 'Age', 'Gender', 'Item ID', 'Item Name', 'Price'], dtype='object')

In [ ]:
### 이곳에서 본인 데이터로 과제를 진행해 주세요 ### 

In [ ]:
# 내가 하고 싶은거
# 1. 내 게임 데이터로 분석할 때
# 1-1. 어떤 캐릭터의 승률이 가장 높은지
# 1-2. 어떤 조건이 내 게임의 승률에 영향을 가장 미치는가?
# 2. 상위 100위 유저 데이터로 분석할 때(각각 최근 100판씩 가져오기)
# 2-1. 어떤 캐릭터의 승률이 가장 높은지
# 2-2. 어떤 조건이 게임의 승률에 가장 큰 영향을 미치는가?

In [ ]:
# 위에는 아직 사이퍼즈 데이터 츄라이즁
# 패키지 로드
import json
import urllib.request, urllib.parse, urllib.error
import ssl
import requests
import pandas as pd

In [ ]:
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

In [ ]:
# request로 데이터 받아오기
# 기본 url: 'https://api.neople.co.kr/cy/{레퍼런스}?apikey={APIKey}&characterName={값}'
# api key: 8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t
url1 = 'https://api.neople.co.kr/cy/players?nickname'
url2 = '&wordType=match&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'
name = "비비두"
url = url1 + urllib.parse.urlencode({'':name}) + url2

url_1 = 'https://api.neople.co.kr/cy/players?nickname=-FURY-&wordType=match&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'
# requestData = requests.get('https://api.neople.co.kr/cy/players?nickname=-FURY-&wordType=match&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t')

In [ ]:
# result = urllib.request.urlopen(url,context=ctx).read()
# info = json.loads(result)

result = urllib.request.urlopen(url,context=ctx).read()
info = json.loads(result)

In [ ]:
info

{'rows': [{'grade': 51,
   'nickname': '비비두',
   'playerId': '5178716601e76cad0d0e2978caf44317'}]}

In [ ]:
playerid = info["rows"][0]["playerId"]
reurl1 = 'https://api.neople.co.kr/cy/players/'
reurl2 = '?apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'

In [ ]:
reurl = reurl1 + playerid + reurl2

In [ ]:
result = urllib.request.urlopen(reurl,context=ctx).read()

In [ ]:
# json 형태로 데이터 로드(유저의 데이터)
info = json.loads(result)

In [ ]:
info

{'clanName': '드러그제너레이션',
 'grade': 51,
 'maxRatingPoint': None,
 'nickname': '비비두',
 'playerId': '5178716601e76cad0d0e2978caf44317',
 'ratingPoint': None,
 'records': [{'gameTypeId': 'rating',
   'loseCount': 0,
   'stopCount': 0,
   'winCount': 0},
  {'gameTypeId': 'normal', 'loseCount': 125, 'stopCount': 1, 'winCount': 117}],
 'tierName': None}

In [ ]:
url = 'https://api.neople.co.kr/cy/players/5178716601e76cad0d0e2978caf44317/matches?gameTypeId=<gameTypeId>&startDate=<startDate>&endDate=<endDate>&limit=<limit>&next=<next>&apikey=<apikey>'

In [ ]:
# 원래는 1~100위의 데이터 각각 최대 100개씩 로드해서 총 1만여건의 데이터를 수집한다! 가 목표였는데
# 추출해본 결과, 8572건밖에 안모여서
# 좀 더 많은 랭커의 데이터를 모으기로 함...이렇게 망겜인걸 알고싶지않았는데
# 랭커도 한달에 100겜도안하다니..........
# 이렇게된거 최대 제한으로 뽑아보기로 함
# 통합 랭킹 1~1000위의 데이터를 로드하자
# 근데 이렇게 해도 68375개밖에 안나옴...흑..흑...어쩔수없다 겜이 망겜인걸 어떡해
# 이 데이터에서 필요한 것: playerId(그래야 경기내역을 조회할 수 있음!)
url = 'https://api.neople.co.kr/cy/ranking/ratingpoint?&limit=1000&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'

In [ ]:
result = urllib.request.urlopen(url,context=ctx).read()

In [ ]:
# json 형태로 데이터 로드(통합 랭킹)
info = json.loads(result)

In [ ]:
info

{'rows': [{'clanName': '전국호타루상향협회',
   'grade': 105,
   'nickname': 'PPIE',
   'playerId': '7963c1386bdbfd6b670a00719f0a6920',
   'rank': 1,
   'ratingPoint': 2624},
  {'clanName': '아싸끼리놀쟈',
   'grade': 104,
   'nickname': '남핀꽃',
   'playerId': '4b4c1cc8b1df1e8f974aace93662e906',
   'rank': 2,
   'ratingPoint': 2600},
  {'clanName': '도리화',
   'grade': 116,
   'nickname': 'ㅡFURYㅡ',
   'playerId': '507a8e01d95536925e76a3d39e3c05b6',
   'rank': 3,
   'ratingPoint': 2565},
  {'clanName': 'JOKERTIM',
   'grade': 88,
   'nickname': '사퍼',
   'playerId': 'a9ecef7c1eadecc604867a71c221dc9c',
   'rank': 4,
   'ratingPoint': 2555},
  {'clanName': '너희다매너해',
   'grade': 111,
   'nickname': '학뀨뀨잉',
   'playerId': '1d876b79f75432619a5d2f93058ab6d2',
   'rank': 5,
   'ratingPoint': 2552},
  {'clanName': '이루리',
   'grade': 71,
   'nickname': '진하',
   'playerId': '80e4ca8a2773daae517ea08583e72876',
   'rank': 6,
   'ratingPoint': 2549},
  {'clanName': 'TearDrop',
   'grade': 101,
   'nickname': '낭만을아시나요'

In [ ]:
# 총 몇 건의 랭킹 데이터가 호출되었는지 확인해보자(정상적으로 불러왔다면 100건임)
print(len(info['rows']))

1000


In [ ]:
playerId = []

In [ ]:
# 차후 playerId로 매칭 기록을 저장해야하므로
# playerId만 따로 저장

for i in range(0, len(info['rows'])):
  playerId.append(info['rows'][i]['playerId'])

In [ ]:
playerId

['7963c1386bdbfd6b670a00719f0a6920',
 '4b4c1cc8b1df1e8f974aace93662e906',
 '507a8e01d95536925e76a3d39e3c05b6',
 'a9ecef7c1eadecc604867a71c221dc9c',
 '1d876b79f75432619a5d2f93058ab6d2',
 '80e4ca8a2773daae517ea08583e72876',
 '244f305537852c50d2fb83cde0f66a6c',
 '7e265f9a658006d06b5c55c4bc517194',
 '31c1a8d06115c217d899ce77dcbaf8f8',
 '10c634e2aefeed9c7221e9edcdf98abc',
 'e9d83c5d9e1d0a529712c63da310b58d',
 '70fec3d93f99ad0268e175953b5612ed',
 'aa2d29d2adbd3d3a8235b3107e7f31a7',
 '5896d9f4d776c05819118882b6258ccc',
 '1e9e8e5df132feb0c2cb925a6beb004a',
 'ba3f3b3095fdaad0c4a02eafa02c8b5d',
 '1505ef03d802f3fac8cd029dc9a9025d',
 '0e090fe998e1d425b925a3c2d2d745c2',
 'b74c1979fbc7432102a6e524ba9a467b',
 '057871f2395fbdb8fe1befd217e7f56d',
 'c049fdcd9f07ce704c73751514cc871d',
 '289d050bfbffbf7c1615de2c6648ef22',
 '4bdfab145f91bf3700e55e2222143631',
 '1d01a2d3c8d9e17757e897591b2cb57d',
 '40d39c23eca38130095cf6408cc3e780',
 'c9d17c580ab5200b011091cf10b78088',
 '6d18f30d22fabc9c4bfc0b62fc62850c',
 

In [ ]:
total_data = []

In [ ]:
test_mapName = []
test_chName = []
test_killCnt = []
test_deathCnt = []
test_assisCnt = []
test_partyCnt = []
test_result = []
test_playT = []
test_spendCoin = []
test_posBuff = []
test_pos = []

In [ ]:
# 이제 playerId 별 데이터를 로드하자
# 공식전 데이터 중
# 2월15일부터 3월14일까지의 데이터 중 최신 100건 로드(최대 100건. 이거보다 더 적을 수도 있음)
reurl1 = 'https://api.neople.co.kr/cy/players/'
reurl2 = '/matches?gameTypeId=rating&startDate=20220215T0000&endDate=20220314T2359&limit=100&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'

for i in playerId:
  reurl = reurl1 + i + reurl2
  result = urllib.request.urlopen(reurl,context=ctx).read()
  # json 형태로 데이터 로드(유저의 매칭 기록)
  info = json.loads(result)

  total_data.append(info)

In [ ]:
len(total_data[5]["matches"]["rows"])

73

In [ ]:
# 이제 playerId 별 데이터를 로드하자
# 공식전 데이터 중
# 2월15일부터 3월14일까지의 데이터 중 최신 100건 로드
reurl1 = 'https://api.neople.co.kr/cy/players/'
reurl2 = '/matches?gameTypeId=rating&startDate=20220215T0000&endDate=20220314T2359&limit=100&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'

for i in playerId:
  reurl = reurl1 + i + reurl2
  result = urllib.request.urlopen(reurl,context=ctx).read()
  # json 형태로 데이터 로드(유저의 매칭 기록)
  info = json.loads(result)

  # 필요한 데이터 추출
  for j in range(0, len(info['matches']['rows'])):
    test_info = info["matches"]["rows"][j]

    test_mapName.append(test_info["map"]["name"])
    test_chName.append(test_info["playInfo"]["characterName"])
    test_killCnt.append(test_info["playInfo"]["killCount"])
    test_deathCnt.append(test_info["playInfo"]["deathCount"])
    test_assisCnt.append(test_info["playInfo"]["assistCount"])
    test_partyCnt.append(test_info["playInfo"]["partyUserCount"])
    test_result.append(test_info["playInfo"]["result"])
    test_playT.append(test_info["playInfo"]["playTime"])
    test_spendCoin.append(test_info["playInfo"]["spendCoin"])
    test_posBuff.append(test_info["position"]["attribute"][0]["name"]+"_"+test_info["position"]["attribute"][1]["name"]+"_"+test_info["position"]["attribute"][2]["name"])
    test_pos.append(test_info["position"]["name"])

In [ ]:
for i in range(0, len(info['matches']['rows'])):
  test_info = info["matches"]["rows"][i]

  test_mapName.append(test_info["map"]["name"])
  test_chName.append(test_info["playInfo"]["characterName"])
  test_killCnt.append(test_info["playInfo"]["killCount"])
  test_deathCnt.append(test_info["playInfo"]["deathCount"])
  test_assisCnt.append(test_info["playInfo"]["assistCount"])
  test_partyCnt.append(test_info["playInfo"]["partyUserCount"])
  test_result.append(test_info["playInfo"]["result"])
  test_playT.append(test_info["playInfo"]["playTime"])
  test_spendCoin.append(test_info["playInfo"]["spendCoin"])
  test_posBuff.append(test_info["position"]["attribute"][0]["name"]+"_"+test_info["position"]["attribute"][1]["name"]+"_"+test_info["position"]["attribute"][2]["name"])
  test_pos.append(test_info["position"]["name"])


In [ ]:
total_data[0]["matches"]["rows"][0]

{'date': '2022-03-14 22:48',
 'map': {'mapId': '103', 'name': '브리스톨'},
 'matchId': '9af506993ba2e27cdd2144743e753dfbef2558ee1a8fae0d0633027d709961e3',
 'playInfo': {'assistCount': 15,
  'attackPoint': 48662,
  'backAttackCount': 40,
  'battlePoint': 145,
  'characterId': '102c0a466e4d5e0e53cdce7f93879d51',
  'characterName': '트리비아',
  'comboCount': 93,
  'damagePoint': 17737,
  'deathCount': 4,
  'demolisherKillCount': 22,
  'getCoin': 17790,
  'guardTowerKillCount': 0,
  'guardianKillCount': 1,
  'healAmount': 0,
  'killCount': 8,
  'level': 58,
  'maxLifeTime': 252,
  'minLifeTime': 13,
  'partyInfo': [],
  'partyUserCount': 0,
  'playTime': 873,
  'playTypeName': '정상',
  'random': False,
  'responseTime': 141,
  'result': 'win',
  'sentinelKillCount': 25,
  'sightPoint': 254,
  'spellCount': 0,
  'spendCoin': 17765,
  'spendConsumablesCoin': 2490,
  'towerAttackPoint': 44552,
  'trooperKillCount': 1},
 'position': {'attribute': [{'id': '71f584223dfa59faa54e3004b7f348be',
    'level'

In [ ]:
# 시작~끝 데이터 로드하는법
#startDate, endDate 요청 변수 사용 예시
  # ex) /matches?startDate=20180901T0000&endDate=20180930T2359
  #         /matches?startDate=2018-09-01 00:00&endDate=2018-09-30 23:59
  # ※ 기간 검색 요청 시 하나의 파라미터(startDate, endDate)라도 누락 될 경우 검색이 불가능합니다.
  # ※ 다양한 Date Type을 지원하지만, 상황에 따라 변경될 수 있으므로 예시 Date Type을 이용 부탁드립니다.
  # ※ 초 단위 설정의 경우 startDate : 00초, endDate : 59초로 처리됩니다.
  # ※ 기간 설정은 최대 90일까지 가능합니다.

In [ ]:
# 매칭 기록 조회
# 일단...가볍게(?) 1만 건만 뽑아보기로 함...내가 그 정도로 게임을 했는지는?모르겠음^^
# 근데 최대 100건씩밖에 조회가 안되서 10일씩 기한을 나눠서 조회해보려 함
# 그리고 전적이 2월 중순에 초기화가 한번 되서^^별수없이 2월~3월 데이터 약 250건으로 내 데이터 테스트
reurl1 = 'https://api.neople.co.kr/cy/players/'
reurl2 = '/matches?gameTypeId=normal&startDate=20220310T0000&endDate=20220314T0359&limit=100&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t'
reurl = reurl1 + playerid + reurl2
result = urllib.request.urlopen(reurl,context=ctx).read()

# json 형태로 데이터 로드(유저의 매칭 기록)
info = json.loads(result)

In [ ]:
info

{'clanName': '드러그제너레이션',
 'grade': 51,
 'matches': {'date': {'end': '2022-03-14 03:59', 'start': '2022-03-10 00:00'},
  'gameTypeId': 'normal',
  'next': None,
  'rows': [{'date': '2022-03-14 02:07',
    'map': {'mapId': '104', 'name': '스프링필드'},
    'matchId': '938a6d68a87080d5606f68fe95d8de7633ece3ee90f426c4e5628d620120d9f4',
    'playInfo': {'assistCount': 12,
     'attackPoint': 33067,
     'backAttackCount': 35,
     'battlePoint': 196,
     'characterId': '3bc12f4fed260b22fa588f22b7f5abff',
     'characterName': '샬럿',
     'comboCount': 42,
     'damagePoint': 24693,
     'deathCount': 3,
     'demolisherKillCount': 28,
     'getCoin': 15490,
     'guardTowerKillCount': 0,
     'guardianKillCount': 1,
     'healAmount': 0,
     'killCount': 4,
     'level': 47,
     'maxLifeTime': 604,
     'minLifeTime': 82,
     'partyInfo': [{'characterId': 'aa7e0636db959e92f01200c5130f19af',
       'characterName': '앨리셔',
       'nickname': 'MAHA',
       'playerId': '574fc0ae6edfdef509a9138f1

In [ ]:
# 총 몇 건의 매칭 데이터가 호출되었는지 확인해보자
print(len(info['matches']['rows']))

44


In [ ]:
play_2028 = len(info['matches']['rows'])

In [ ]:
play_1019 = len(info['matches']['rows'])

In [ ]:
play_0109 = len(info['matches']['rows'])

In [ ]:
play_030109 = len(info['matches']['rows'])

In [ ]:
play_031014 = len(info['matches']['rows'])

In [ ]:
play_2028 + play_1019 + play_030109 + play_031014

247

In [ ]:
# 우리가 필요한 데이터는 matches에 들어있는 정보이므로,
# 뭘 뽑아내야 하는지 알아보기 위해 매칭데이터 하나만 뽑아내보았다
info["matches"]["rows"][0]

{'date': '2022-03-14 01:45',
 'map': {'mapId': '102', 'name': '메트로폴리스'},
 'matchId': '39eb368b645af9e1760fefe439a36d1043235ecbdf4f07bf2b154e2e88ac1c95',
 'playInfo': {'assistCount': 6,
  'attackPoint': 15108,
  'backAttackCount': 8,
  'battlePoint': 45,
  'characterId': '8ca68ab00e6f2c33220214d720c024ee',
  'characterName': '에밀리',
  'comboCount': 33,
  'damagePoint': 26445,
  'deathCount': 7,
  'demolisherKillCount': 32,
  'getCoin': 14460,
  'guardTowerKillCount': 0,
  'guardianKillCount': 0,
  'healAmount': 0,
  'killCount': 1,
  'level': 49,
  'maxLifeTime': 259,
  'minLifeTime': 43,
  'partyInfo': [{'characterId': '4408e80fb00f3e4a1e1f27247ca433c9',
    'characterName': '티샤',
    'nickname': 'MAHA',
    'playerId': '574fc0ae6edfdef509a9138f166aa151'}],
  'partyUserCount': 2,
  'playTime': 1094,
  'playTypeName': '정상',
  'random': False,
  'responseTime': 231,
  'result': 'lose',
  'sentinelKillCount': 16,
  'sightPoint': 82,
  'spellCount': 0,
  'spendCoin': 13725,
  'spendConsumab

In [ ]:
# map - 맵 정보. map 중에서도 name을 뽑아내야 한다.
# 형태: 'map': {'mapId': '102', 'name': '메트로폴리스'}


# ---------이 이하에 있는건 'playInfo' 안에 들어있음-----------------
# characterName: 플레이 한 캐릭터
# 형태: 'characterName': '에밀리'

# KDA를 계산해서 넣으면 좋을 것 같음
# 이를 위해 뽑아내야 할 데이터: 킬수, 데스수, 어시스트수
# killCount: 킬 수
# 형태: 'killCount': 1
# deathCount: 데스 수
# 형태: 'deathCount': 7
# assistCount: 어시스트 수
# 형태: 'assistCount': 6

# 파티원의 수와 파티여부를 만드는것도 좋을 것 같음
# 만약 partyUserCount가 0 이면 N, 1 이상이면 Y를 표시하도록 하자
# 이것도 추후 행으로 추가하도록 하고 일단 파티원수 데이터를 뽑아내기로 함
# partyUserCount: 파티원수
# 형태: 'partyUserCount': 2

# result: 승리여부 <- 이게 우리의 타겟변수가 될 예정
# 형태: 'result': 'lose' / 'result': 'win'

# playTime: 플레이시간
# 형태: 'playTime': 1094

# spendCoin: 코인소모량
# 형태: 'spendCoin': 13725

# ---------이 이하에 있는건 'position' 안에 들어있음-----------------
# 포지션 버프는 여러개 있지만
# 어떻게 3개를 조합하는게 승률에 영향을 주는지 알아보려 함
# 따라서, position안의 attribute에 들어있는 포지션버프이름인 name을 뽑아
# 하나의 string으로 만들어 컬럼으로 저장하고싶음
# 예를 들어, '전장의 열기-할인 판매-태세전환' 이런식으로
# level이 1인것의 name을 찾아서 저장해두고 2인걸 찾아서 저장해두고 3인걸 찾아서 저장해둔다음에
# 이런형태로 합쳐서 저장

# naem: 포지션명
# 형태: 'name': '원거리딜러'

# 만약 성공적으로 모두 만들어진다면,
# 1개의 타겟과 약 10개의 피처로 이루어진 데이터가 만들어짐

In [ ]:
# 테스트 추출
first_info = info["matches"]["rows"][0]

test_mapName = first_info["map"]["name"]
test_chName = first_info["playInfo"]["characterName"]
test_killCnt = first_info["playInfo"]["killCount"]
test_deathCnt = first_info["playInfo"]["deathCount"]
test_assisCnt = first_info["playInfo"]["assistCount"]
test_partyCnt = first_info["playInfo"]["partyUserCount"]
test_result = first_info["playInfo"]["result"]
test_playT = first_info["playInfo"]["playTime"]
test_spendCoin = first_info["playInfo"]["spendCoin"]
test_posBuff = first_info["position"]["attribute"][0]["name"]+"_"+first_info["position"]["attribute"][1]["name"]+"_"+first_info["position"]["attribute"][2]["name"]
test_pos = first_info["position"]["name"]

In [ ]:
test_mapName

'메트로폴리스'

In [ ]:
test_chName

'에밀리'

In [ ]:
test_killCnt

1

In [ ]:
test_deathCnt

[3,
 7,
 6,
 8,
 2,
 5,
 5,
 4,
 6,
 7,
 6,
 4,
 7,
 7,
 11,
 9,
 5,
 5,
 5,
 2,
 4,
 5,
 4,
 2,
 3,
 9,
 7,
 7,
 7,
 8,
 5,
 10,
 6,
 7,
 10,
 3,
 6,
 8,
 3,
 9,
 7,
 4,
 2,
 5]

In [ ]:
test_assisCnt

6

In [ ]:
test_partyCnt

[2,
 2,
 2,
 3,
 3,
 3,
 4,
 4,
 4,
 4,
 4,
 4,
 3,
 3,
 3,
 5,
 5,
 5,
 4,
 3,
 3,
 3,
 3,
 3,
 3,
 4,
 3,
 3,
 3,
 2,
 2,
 3,
 3,
 3,
 3,
 3,
 0,
 0,
 0,
 3,
 3,
 3,
 3,
 0,
 2,
 4,
 4,
 4,
 2,
 2,
 2,
 2,
 3,
 5,
 5,
 5,
 5,
 5,
 4,
 3,
 3,
 3,
 3,
 4,
 4,
 4,
 3,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 2,
 2,
 2,
 2,
 2,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 2,
 2,
 2,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 3,
 3,
 3,
 3,
 0,
 0,
 0,
 0,
 0,
 0,
 4,
 4,
 4,
 4,
 4,
 2,
 2,
 2,
 2,
 3,
 3,
 3,
 3]

In [ ]:
test_result

'lose'

In [ ]:
test_playT

1094

In [ ]:
test_spendCoin

13725

In [ ]:
test_posBuff

'전장의 열기_할인 판매_태세전환'

In [ ]:
test_pos

'원거리딜러'

In [ ]:
test_mapName = []
test_chName = []
test_killCnt = []
test_deathCnt = []
test_assisCnt = []
test_partyCnt = []
test_result = []
test_playT = []
test_spendCoin = []
test_posBuff = []
test_pos = []

In [ ]:
len(test_mapName)

68375

In [ ]:
len(test_mapName)
len(test_chName)
len(test_killCnt)
len(test_deathCnt)
len(test_assisCnt)
len(test_partyCnt)
len(test_result)
len(test_playT)
len(test_spendCoin)
len(test_posBuff)
len(test_pos)

68375

In [ ]:
for i in range(0, 5):
  print(i)

0
1
2
3
4


In [ ]:
# 일단 오늘은 여기까지하고
# 내일 구현할 알고리즘
# len(info['matches']['rows']) 만큼 반복해서
# 위처럼 데이터를 추출한 뒤
# 이를 뭐 어디 저장해서
# 컬럼에 쌓기


# 빈 리스트 생성(여기에 데이터를 담을 예정)

# test_mapName = first_info["map"]["name"]
# test_chName = first_info["playInfo"]["characterName"]
# test_killCnt = first_info["playInfo"]["killCount"]
# test_deathCnt = first_info["playInfo"]["deathCount"]
# test_assisCnt = first_info["playInfo"]["assistCount"]
# test_partyCnt = first_info["playInfo"]["partyUserCount"]
# test_result = first_info["playInfo"]["result"]
# test_playT = first_info["playInfo"]["playTime"]
# test_spendCoin = first_info["playInfo"]["spendCoin"]
# test_posBuff = first_info["position"]["attribute"][0]["name"]+"_"+first_info["position"]["attribute"][1]["name"]+"_"+first_info["position"]["attribute"][2]["name"]
# test_pos = first_info["position"]["name"]


for i in range(0, len(info['matches']['rows'])):
  test_info = info["matches"]["rows"][i]

  test_mapName.append(test_info["map"]["name"])
  test_chName.append(test_info["playInfo"]["characterName"])
  test_killCnt.append(test_info["playInfo"]["killCount"])
  test_deathCnt.append(test_info["playInfo"]["deathCount"])
  test_assisCnt.append(test_info["playInfo"]["assistCount"])
  test_partyCnt.append(test_info["playInfo"]["partyUserCount"])
  test_result.append(test_info["playInfo"]["result"])
  test_playT.append(test_info["playInfo"]["playTime"])
  test_spendCoin.append(test_info["playInfo"]["spendCoin"])
  test_posBuff.append(test_info["position"]["attribute"][0]["name"]+"_"+test_info["position"]["attribute"][1]["name"]+"_"+test_info["position"]["attribute"][2]["name"])
  test_pos.append(test_info["position"]["name"])






# 그렇게 내 데이터로 한번 분석 실행해보고
# 다음번엔 상위랭커 100명의 공식전 데이터 추출

In [ ]:
# 리스트를 dataframe으로 변형

cname = ['mapName', 'character', 'kill', 'death', 'assist', 'partyCnt', 'result', 'playTime', 'spendCoin', 'posBuff', 'possition']

df = pd.DataFrame(zip(test_mapName, test_chName, test_killCnt, test_deathCnt, test_assisCnt, test_partyCnt, test_result, test_playT, test_spendCoin, test_posBuff, test_pos), columns = cname)

df

mapName character  kill  death  assist  partyCnt result  playTime  \
0             브리스톨      트리비아     8      4      15         0    win       873   
1      그랑플람 아시아 지부      트리비아    13      2      12         0   lose      1212   
2             리버포드      트리비아     8      6      12         0   lose      1394   
3      그랑플람 아시아 지부      트리비아     8      4      10         0    win       753   
4           메트로폴리스        리사     0      3      28         2   lose       991   
...            ...       ...   ...    ...     ...       ...    ...       ...   
68370  그랑플람 아시아 지부        휴톤     5     11       0         0   lose      1147   
68371         리버포드       케니스     4      5       7         0    win      1012   
68372       메트로폴리스       헤나투     5     10       4         0   lose      1194   
68373        스프링필드       윌라드     2      4       9         0    win       865   
68374         리버포드        시바     5     10       9         0    win      1146   

       spendCoin               posBuff possition  
0          17765    충동 구매_할인 판매_전장의 열기     원거리딜러  
1          21475    전투 태세_할인 판매_불굴의 의지        탱커  
2          21675    충동 구매_할인 판매_전장의 열기     원거리딜러  
3          17015    충동 구매_할인 판매_전장의 열기     원거리딜러  
4          11700   근육 손상_치유 증폭기_활성화 지원       서포터  
...          ...                   ...       ...  
68370      14400  불굴의 의지_재생의 숨결_최후의 저항        탱커  
68371      17460    급소 가격_할인 판매_거인 사냥꾼     근거리딜러  
68372      15920     태세전환_할인 판매_전장의 열기     원거리딜러  
68373      14550     태세전환_할인 판매_전장의 열기     원거리딜러  
68374      17890    급소 가격_할인 판매_거인 사냥꾼     근거리딜러  

[68375 rows x 11 columns]

In [ ]:
# 파일로 저장
df.to_csv('rankdata_220215_220314.csv', index = False, encoding = 'CP949')

In [ ]:
url = "https://api.neople.co.kr/cy/matches/3332d79b94bf2fa265650301dda2f57cdde0c3f430bf7e7bf8854c54c913be60?&apikey=8w5jBaG23mzTDfOI2UO29XVUH1JQVs2t"

In [ ]:
result = urllib.request.urlopen(url,context=ctx).read()

# json 형태로 데이터 로드(매칭 상세 정보)
info = json.loads(result)

In [ ]:
info

{'date': '2022-03-13 18:37',
 'gameTypeId': 'normal',
 'map': {'mapId': '101', 'name': '리버포드'},
 'players': [{'items': [{'equipSlotCode': '101',
     'equipSlotName': '손(공격)',
     'itemId': '07365a6e93c3bce23fde80cd3b56f9cc',
     'itemName': 'E 트론 룰러 ',
     'rarityCode': '104',
     'rarityName': '유니크',
     'slotCode': '101',
     'slotName': '손(공격)'},
    {'equipSlotCode': '102',
     'equipSlotName': '머리(치명)',
     'itemId': '59fd2edb41ac72289389861624b02a5a',
     'itemName': '핀 플러그 ',
     'rarityCode': '103',
     'rarityName': '레어',
     'slotCode': '102',
     'slotName': '머리(치명)'},
    {'equipSlotCode': '103',
     'equipSlotName': '가슴(체력)',
     'itemId': '612f25190b4fb3eb2eb9e55539db9f6a',
     'itemName': 'E 랜드 스피더',
     'rarityCode': '104',
     'rarityName': '유니크',
     'slotCode': '103',
     'slotName': '가슴(체력)'},
    {'equipSlotCode': '104',
     'equipSlotName': '허리(회피)',
     'itemId': '9321c1245d595bac53103f0faf7fbbc5',
     'itemName': 'E 퓨처 워커',
     'rarityCo

In [ ]:
# 나머지는 차후 자세히 랭글링하기로!

In [ ]:
# 2. 포켓몬 데이터 랭글링
# 구글 드라이브에서 데이터 로드(kaggle에서 데이터를 다운받은 뒤 구글드라이브에 저장하였다)
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
file_name = '/content/drive/MyDrive/n23/pokemon.csv'

pk_df = pd.read_csv(file_name)

In [ ]:
pk_df.head(40)

abilities  against_bug  \
0                         ['Overgrow', 'Chlorophyll']         1.00   
1                         ['Overgrow', 'Chlorophyll']         1.00   
2                         ['Overgrow', 'Chlorophyll']         1.00   
3                            ['Blaze', 'Solar Power']         0.50   
4                            ['Blaze', 'Solar Power']         0.50   
5                            ['Blaze', 'Solar Power']         0.25   
6                            ['Torrent', 'Rain Dish']         1.00   
7                            ['Torrent', 'Rain Dish']         1.00   
8                            ['Torrent', 'Rain Dish']         1.00   
9                         ['Shield Dust', 'Run Away']         1.00   
10                                      ['Shed Skin']         1.00   
11                    ['Compoundeyes', 'Tinted Lens']         0.50   
12                        ['Shield Dust', 'Run Away']         0.50   
13                                      ['Shed Skin']         0.50   
14                                ['Swarm', 'Sniper']         0.50   
15          ['Keen Eye', 'Tangled Feet', 'Big Pecks']         0.50   
16          ['Keen Eye', 'Tangled Feet', 'Big Pecks']         0.50   
17          ['Keen Eye', 'Tangled Feet', 'Big Pecks']         0.50   
18  ['Run Away', 'Guts', 'Hustle', 'Gluttony', 'Hu...         1.00   
19  ['Run Away', 'Guts', 'Hustle', 'Gluttony', 'Hu...         1.00   
20                             ['Keen Eye', 'Sniper']         0.50   
21                             ['Keen Eye', 'Sniper']         0.50   
22             ['Intimidate', 'Shed Skin', 'Unnerve']         0.50   
23             ['Intimidate', 'Shed Skin', 'Unnerve']         0.50   
24                         ['Static', 'Lightningrod']         1.00   
25         ['Static', 'Lightningrod', 'Surge Surfer']         1.00   
26  ['Sand Veil', 'Sand Rush', 'Snow Cloak', 'Slus...         1.00   
27  ['Sand Veil', 'Sand Rush', 'Snow Cloak', 'Slus...         1.00   
28              ['Poison Point', 'Rivalry', 'Hustle']         0.50   
29              ['Poison Point', 'Rivalry', 'Hustle']         0.50   
30         ['Poison Point', 'Rivalry', 'Sheer Force']         0.50   
31              ['Poison Point', 'Rivalry', 'Hustle']         0.50   
32              ['Poison Point', 'Rivalry', 'Hustle']         0.50   
33         ['Poison Point', 'Rivalry', 'Sheer Force']         0.50   
34      ['Cute Charm', 'Magic Guard', 'Friend Guard']         0.50   
35           ['Cute Charm', 'Magic Guard', 'Unaware']         0.50   
36  ['Flash Fire', 'Drought', 'Snow Cloak', 'Snow ...         0.50   
37  ['Flash Fire', 'Drought', 'Snow Cloak', 'Snow ...         0.50   
38      ['Cute Charm', 'Competitive', 'Friend Guard']         0.50   
39             ['Cute Charm', 'Competitive', 'Frisk']         0.50   

    against_dark  against_dragon  against_electric  against_fairy  \
0            1.0             1.0               0.5            0.5   
1            1.0             1.0               0.5            0.5   
2            1.0             1.0               0.5            0.5   
3            1.0             1.0               1.0            0.5   
4            1.0             1.0               1.0            0.5   
5            1.0             1.0               2.0            0.5   
6            1.0             1.0               2.0            1.0   
7            1.0             1.0               2.0            1.0   
8            1.0             1.0               2.0            1.0   
9            1.0             1.0               1.0            1.0   
10           1.0             1.0               1.0            1.0   
11           1.0             1.0               2.0            1.0   
12           1.0             1.0               1.0            0.5   
13           1.0             1.0               1.0            0.5   
14           1.0             1.0               1.0            0.5   
15           1.0             1.0               2.0            1.0   


In [ ]:
pk_df.columns

Index(['abilities', 'against_bug', 'against_dark', 'against_dragon',
       'against_electric', 'against_fairy', 'against_fight', 'against_fire',
       'against_flying', 'against_ghost', 'against_grass', 'against_ground',
       'against_ice', 'against_normal', 'against_poison', 'against_psychic',
       'against_rock', 'against_steel', 'against_water', 'attack',
       'base_egg_steps', 'base_happiness', 'base_total', 'capture_rate',
       'classfication', 'defense', 'experience_growth', 'height_m', 'hp',
       'japanese_name', 'name', 'percentage_male', 'pokedex_number',
       'sp_attack', 'sp_defense', 'speed', 'type1', 'type2', 'weight_kg',
       'generation', 'is_legendary'],
      dtype='object')

In [ ]:
# 1. 가장 많은 타입은?
# 그런데 타입이 2개인 애들도 있어서
# type 1과 2를 확인해보려 함
# 1-1. type이 1개만 있는 애들 구분하기.
pk_df['type2'].isnull().sum()

384

In [ ]:
pk_df['type1'].isnull().sum()

0

In [ ]:
# 타입이 1개인 애들은 384개. 2개인 애들은 417개
# 오..타입이 2개인 애들과 하나인 애들끼리 비교해도 좋을거같다

In [ ]:
pk_type1 = pk_df[pk_df['type2'].isnull()]
pk_type1

abilities  against_bug  against_dark  \
3                     ['Blaze', 'Solar Power']          0.5           1.0   
4                     ['Blaze', 'Solar Power']          0.5           1.0   
6                     ['Torrent', 'Rain Dish']          1.0           1.0   
7                     ['Torrent', 'Rain Dish']          1.0           1.0   
8                     ['Torrent', 'Rain Dish']          1.0           1.0   
..                                         ...          ...           ...   
781  ['Bulletproof', 'Soundproof', 'Overcoat']          1.0           1.0   
788                                ['Unaware']          2.0           2.0   
789                                 ['Sturdy']          2.0           2.0   
795                            ['Beast Boost']          1.0           1.0   
799                            ['Prism Armor']          2.0           2.0   

     against_dragon  against_electric  against_fairy  against_fight  \
3               1.0               1.0            0.5            1.0   
4               1.0               1.0            0.5            1.0   
6               1.0               2.0            1.0            1.0   
7               1.0               2.0            1.0            1.0   
8               1.0               2.0            1.0            1.0   
..              ...               ...            ...            ...   
781             2.0               0.5            2.0            1.0   
788             1.0               1.0            1.0            0.5   
789             1.0               1.0            1.0            0.5   
795             1.0               0.5            1.0            1.0   
799             1.0               1.0            1.0            0.5   

     against_fire  against_flying  against_ghost  ...  percentage_male  \
3             0.5             1.0            1.0  ...             88.1   
4             0.5             1.0            1.0  ...             88.1   
6             0.5             1.0            1.0  ...             88.1   
7             0.5             1.0            1.0  ...             88.1   
8             0.5             1.0            1.0  ...             88.1   
..            ...             ...            ...  ...              ...   
781           0.5             1.0            1.0  ...             50.0   
788           1.0             1.0            2.0  ...              NaN   
789           1.0             1.0            2.0  ...              NaN   
795           1.0             0.5            1.0  ...              NaN   
799           1.0             1.0            2.0  ...              NaN   

     pokedex_number  sp_attack  sp_defense  speed     type1  type2  weight_kg  \
3                 4         60          50     65      fire    NaN        8.5   
4                 5         80          65     80      fire    NaN       19.0   
6                 7         50          64     43     water    NaN        9.0   
7                 8         65          80     58     water    NaN       22.5   
8                 9        135         115     78     water    NaN       85.5   
..              ...        ...         ...    ...       ...    ...        ...   
781             782         45          45     45    dragon    NaN       29.7   
788             789         29          31     37   psychic    NaN        0.1   
789             790         29         131     37   psychic    NaN      999.9   
795             796        173          71     83  electric    NaN      100.0   
799             800        127          89     79   psychic    NaN      230.0   

     generation  is_legendary  
3             1             0  
4             1             0  
6             1             0  
7             1             0  
8             1             0  
..          ...           ...  
781           7             0  
788           7             1  
789           7             1  
795           7             1  
799           7             1  

[384 rows x 

In [ ]:
pk_type2 = pk_df[pk_df['type2'].notna()]
pk_type2

abilities  against_bug  against_dark  \
0        ['Overgrow', 'Chlorophyll']         1.00           1.0   
1        ['Overgrow', 'Chlorophyll']         1.00           1.0   
2        ['Overgrow', 'Chlorophyll']         1.00           1.0   
5           ['Blaze', 'Solar Power']         0.25           1.0   
11   ['Compoundeyes', 'Tinted Lens']         0.50           1.0   
..                               ...          ...           ...   
794                  ['Beast Boost']         0.50           0.5   
796                  ['Beast Boost']         0.25           1.0   
797                  ['Beast Boost']         1.00           1.0   
798                  ['Beast Boost']         2.00           0.5   
800                   ['Soul-Heart']         0.25           0.5   

     against_dragon  against_electric  against_fairy  against_fight  \
0               1.0               0.5            0.5           0.50   
1               1.0               0.5            0.5           0.50   
2               1.0               0.5            0.5           0.50   
5               1.0               2.0            0.5           0.50   
11              1.0               2.0            1.0           0.25   
..              ...               ...            ...            ...   
794             1.0               1.0            2.0           0.50   
796             0.5               2.0            0.5           1.00   
797             0.5               0.5            0.5           2.00   
798             2.0               0.5            4.0           2.00   
800             0.0               1.0            0.5           1.00   

     against_fire  against_flying  against_ghost  ...  percentage_male  \
0             2.0             2.0            1.0  ...             88.1   
1             2.0             2.0            1.0  ...             88.1   
2             2.0             2.0            1.0  ...             88.1   
5             0.5             1.0            1.0  ...             88.1   
11            2.0             2.0            1.0  ...             50.0   
..            ...             ...            ...  ...              ...   
794           2.0             4.0            1.0  ...              NaN   
796           2.0             0.5            1.0  ...              NaN   
797           4.0             1.0            1.0  ...              NaN   
798           0.5             1.0            0.5  ...              NaN   
800           2.0             0.5            1.0  ...              NaN   

     pokedex_number  sp_attack  sp_defense  speed  type1     type2  weight_kg  \
0                 1         65          65     45  grass    poison        6.9   
1                 2         80          80     60  grass    poison       13.0   
2                 3        122         120     80  grass    poison      100.0   
5                 6        159         115    100   fire    flying       90.5   
11               12         90          80     70    bug    flying       32.0   
..              ...        ...         ...    ...    ...       ...        ...   
794             795        137          37    151    bug  fighting       25.0   
796             797        107         101     61  steel    flying      999.9   
797             798         59          31    109  grass     steel        0.1   
798             799         97          53     43   dark    dragon      888.0   
800             801        130         115     65  steel     fairy       80.5   

     generation  is_legendary  
0             1             0  
1             1             0  
2             1             0  
5             1             0  
11            1             0  
..          ...           ...  
794           7             1  
796           7             1  
797           7             1  
798           7             1  
800           7             1  

[417 rows x 41 columns]

In [ ]:
# 타입이 하나인 애들 중 
# 1. 가장 많은 타입은?
pk_type1['type1'].value_counts()
# 노말타입, 물타입이 각각 61개씩으로 가장 많다

normal      61
water       61
grass       37
psychic     35
fire        27
electric    26
fighting    22
bug         18
fairy       16
poison      13
dragon      12
ice         12
rock        11
ground      10
dark         9
ghost        9
steel        4
flying       1
Name: type1, dtype: int64

In [ ]:
# 타입이 2개인 애들 중 
# 1. 가장 많은 타입은?
pk_type2['type1'].value_counts()
# 벌레타입이 가장 많았다

bug         54
water       53
normal      44
grass       41
rock        34
fire        25
ground      22
dark        20
steel       20
poison      19
ghost       18
psychic     18
dragon      15
electric    13
ice         11
fighting     6
fairy        2
flying       2
Name: type1, dtype: int64

In [ ]:
pk_type2['type2'].value_counts()
# 2번째 타입으로는 비행타입이 제일 많았다

flying      95
poison      34
ground      34
fairy       29
psychic     29
fighting    25
steel       22
dark        21
grass       20
water       17
dragon      17
ice         15
rock        14
ghost       14
fire        13
electric     9
bug          5
normal       4
Name: type2, dtype: int64

### 2) 새로 특징들을 만들어 보세요
- **(Urclass Quiz) purchase_data를 랭글링하여 다음 조건을 계산하고 결과가 올바른 항목을 고르세요.**
    1. 나이 1-20세 그룹(20세 포함)의 평균 Price는 3.101818이다.
    2. 나이 21-50세 그룹의 평균 Price는 3.021559이다.
    3. 가장 많이 구입한 아이템의 ID는 178 이다.
    4. 두번째로 매출을 많이 낸 아이템 ID는 145이다.

In [ ]:
### 이곳에서 Urlcass 퀴즈(assignment)과제를 진행해 주세요 ###
# 그룹별 Price 평균
print(df.Price[df['Age'] <= 20].mean())
# 이거 일단 50포함으로 계산해봄
print(df.Price[(df['Age'] > 20) & (df['Age'] <= 50)].mean())

# 판매수 상위 Item ID (제일 위에꺼)
print(df['Item ID'].mode().loc[0])
# 매출 상위 Item IDs
print(df.groupby('Item ID')['Price'].sum().sort_values(ascending=False).reset_index().iloc[1,0])

3.101818181818184
3.0215587044534415
178
82


In [ ]:
### 이곳에서 본인 데이터로 과제를 진행해 주세요 ### 

## 🔥 도전과제(Github - Discussion)

### 3) 데이터 랭글링 방법들 중 가장 중요하다고 생각하는 방법이나 어려운 점, 혹은 수업에서 배우지 않은 좋은 방법이나 사례가 있으면 서로 공유해 보세요. 


In [ ]:
### 이곳에서 과제를 진행해 주세요 ### 